### PR TITLE
[Proof of Concept/Documentation - Do not merge] Format Parquet files on Azure Snapshot Export

### PR DESCRIPTION
**Pull request by snf2ye**
https://broadworkbench.atlassian.net/browse/DC-1169
__________

# Background - TDR & Parquet Files
Parquet Format POC

### Parquet files
Parquet files are special in that they contain both a schema and tabular data. This schema can be used in place of the TDR snapshot schema to determine the type of data for each column on export to workspace. 

# The Problem: The Parquet Schema generated by Synapse

When we generate parquet files with Synapse, we rely on the data type translations between TSQL and the parquet data types. Synapse translates "datetime2" to "INT96" . This is a deprecated data type - everywhere we export parquet data to expects a long/timestamp data type instead.  We also run into problems with other data types, most importantly arrays, but this proof of concept focuses on datetimes (More details in this [doc](https://docs.google.com/document/d/1fLfYUXfmnZjawq6NqvSHicq0RiGuoXSZ5TcUoud4AEA/edit#heading=h.286j2qnwxshg)). 

From ([Azure Documentation](https://learn.microsoft.com/en-us/azure/synapse-analytics/sql/develop-openrowset#type-mapping-for-parquet)):
image

**Why do we care?**
- Within TDR, we only read parquet data through synapse. And synapse knows how to deal with int96 data types, so it isn't a problem within TDR. We need synapse in order to utilize the SQL querying ability. 
- However, when we go to export TDR data into Terra workspaces, they would have to add special handling for Azure parquet files. Terra workspaces already know how to deal with GCP exported parquet files. So, ideally, we want the  parquet files' schema exported from Azure to match the parquet file schema exported from GCP. This eliminates the need for special handling of the files per cloud. 

> GOAL: Parquet files schema exported from Azure to match the parquet file schema exported from GCP.

**Target: Match the data types produced on Big Query export to parquet**

image

Target of Proof of Concept: Can we re-format parquet files to switch INT96 to timestamp-micros?

# Proof of Concept Approach

Utilize the Avro/Parquet reader/writer java client to stream files from azure and then write the formatted files back to azure.

The goal of this proof of concept was to prove that we could accomplish the following items:
 ✅ Read directly from azure storage blob
 ✅ Write directly to azure storage blob
 ✅ Could reformat the records with a new schema

# Potential Implementation Approach
### Option 1 - On first export of snapshot, generate correctly formatted files

Add steps to snapshot export:
1) Check if the correctly formatted parquet files already exist. If they exist, skip the next steps.
2) Generate sas token for the snapshot parquet file folder
3) Build a parquet schema based on the snapshot schema in TDR using the Avro Schema Builder
4) Follow a similar pattern to the e2e test to stream records through our service reading the source parquet files and writing the correctly formatted parquet files to a new folder
Parquet Format POC (1)

### Option 2 - Generate correctly formatted files on Snapshot Create

Add steps to snapshot create:
After parquet files are created - (SAS tokens should already exist)
1) Build a parquet schema based on the snapshot schema in TDR using the Avro Schema Builder
2) Follow a similar pattern to the e2e test to stream records through our service reading the source parquet files and writing the correctly formatted parquet files to a new folder

Parquet Format POC (2)

For both of these options, we need to instead point our retrieve snapshot and export snapshot manifest to point to the new folder with the newly formatted parquet files instead of the original parquet files.

# Potential concerns
* We are serially walking through each record - How slow would this be for large snapshots? In theory, each parquet file should not be too large to actually break down, but it could take a while to process each file
* Storing copy of the data. But, should be pretty small since parquet files are compressed. 
* We should test more fine grained datetime examples to confirm that the conversion from the byte array -> microseconds since epoch correctly translates. Could any information be lost in the conversion?

### CODE REVIEW

File: src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+  public static void writeToParquet(
+      List<GenericData.Record> recordsToWrite,
+      List<String> columnNames,
+      Map<String, TableDataType> columnDataTypeMap,
+      OutputFile fileToWrite,
+      Schema schema,
+      Configuration config)
+      throws IOException {
+    try (ParquetWriter<GenericData.Record> writer =
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java:59]: It looks like you could open both a writer and reader here and then stream the data without having to read it all into memory, is that what you were thinking we'd do?

File: src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+              if (dtFixed == null) {
+                newRecord.put(column, null);
+              } else {
+                var bytes = dtFixed.bytes();
+                ByteBuffer bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN);
+                long timeOfDayNanos = bb.getLong();
+                var julianDay = bb.getInt();
+                // Given timeOfDayNanos and julianDay, convert to microseconds since epoch
+                Long microSeconds =
+                    (long) (julianDay - 2440588) * 86400 * 1000000 + timeOfDayNanos / 1000;
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java]: You'll want to include a link to the source of this magic formula. Not sure where you got it from, I found it here: https://github.com/xhochy/parquet-format/blob/cb4727767823ae201fd567f67825cc22834c20e9/LogicalTypes.md#int96-timestamps-also-called-impala_timestamp

Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java]: Another reference: https://github.com/dremio/dremio-oss/blob/master/sabot/kernel/src/main/java/com/dremio/exec/store/parquet/ParquetReaderUtility.java#L70

File: src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java (hunk 1 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java:59]: That's a great point - I did not think of it! I just made this improvement to the e2e example.

File: src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java (hunk 2 above)
Comment (snf2ye) [reply src/main/java/bio/terra/service/parquet/ParquetReaderWriterWithAvro.java]: I've included the link in the code comments. Thanks for tracking down that first reference!

**Pull request closed without merging**